### PR TITLE
Live Demo Multitenancy setup

### PR DESCRIPTION
**Pull request by mathemancer**
Related to #1906 

This sets up "invisible" multitenancy for our Live Demo. With this setup, each new session that's connected should receive its own separate copy of the test data sets (for now, just the Library Management data set).

**Technical details**

The multitenancy is set up off of the session ID. This means that connecting from different browsers will result in separate data sets. Same for private windows in the same browser (under most circumstances).
- This should _not_ be considered a secure way to keep data separate.
- This was the quickest way to get things working without implementing true multitenancy.
- A user that connects to the root address will be rerouted to a database with a key appended (the key is deterministically derived from their `sessionid` cookie once they're logged in).

The way database reflection works had to be changed. This change would be necessary anyway in the long run, so should be kept. The difference is we no longer "reflect" database information from the `config/settings.py` file on-the-fly. This is not recommended in the first place, and should have been avoided in the first place. Instead, we reflect database info on the first load of the App, and proceed from there. New databases can be (and are) added via the `Database` model.

### TODO

- Some tests made incorrect assumptions about how to handle multiple databases. Those tests had to be skipped for now.
- The new logic is largely untested.
  - I have more faith in people clicking around, given the current state of our code base.
  - I will add some tests for this behavior if I get time, but it's lower priority given the temporary nature of this solution.
- This may increase problems with state from `metadata`. I didn't notice any with my testing, but I can only test one `sessionid` at a time (for quick requests, anyway). A follow up PR will improve the metadata state handling in a way that should help.

## Checklist
 
- [X] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [X] My pull request targets the `master` branch of the repository
- [X] My commit messages follow [best practices][best_practices].
- [X] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [X] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: config/settings.py
Comment (kgodey) [inline config/settings.py:93]: I'd recommend namespacing settings that are specific to Mathesar i.e. `MATHESAR_LIVE_DEMO`, since we may install other Django apps with generic setting names.

File: config/settings.py
Comment (kgodey) [inline config/settings.py:94]: Is this a default Django setting? If not, we should namespace it too.

File: db/install.py
Comment (kgodey) [inline db/install.py:19]: Does this fix https://github.com/centerofci/mathesar/issues/1481?

File: mathesar/api/dj_filters.py
Comment (silentninja) [inline mathesar/api/dj_filters.py:80]: Why are thse filters needed for multi-tenancy?

File: mathesar/tests/api/test_schema_api.py
Comment (kgodey) [inline mathesar/tests/api/test_schema_api.py:53]: Please create an issue to fix and re-enable these at some point, might be a good `help wanted` issue.

**Review by kgodey — COMMENTED** [review]
I'll let @silentninja take a full look at this, I just skimmed through the code and added some comments. 

I'm wondering if we need the DB-per-session functionality. Could we maintain a list of DBs in Ansible or in some script elsewhere and figure out which DB to send people to _outside_ of the product? I'm mainly trying to avoid building a bunch of live demo related middleware or conditionals directly in the product code. 

If this will take too long to figure out, we can create an issue for it instead of fixing it now.

File: mathesar/tests/api/test_schema_api.py
Comment (mathemancer) [reply mathesar/tests/api/test_schema_api.py:53]: We need to rework the whole testing methodology around DBs. These tests being invalid is a symptom of an underlying problem. I'll make an issue for that, though.

File: db/install.py
Comment (mathemancer) [reply db/install.py:19]: Yes, actually.

File: config/settings.py
Comment (mathemancer) [reply config/settings.py:94]: This appears to be set up by `pytest-django` automatically when we run the tests.

File: config/settings.py
Comment (mathemancer) [reply config/settings.py:93]: Good point, will do.

**Review by silentninja — APPROVED** [review]
Nice work @mathemancer I didn't come across any bugs when testing out the PR locally. But there few issues that make me concerned

1. Leaking staging-related code into the product codebase - Are you planning to merge this PR into the master? If so we should think of abstracting staging-related code(maybe starting with a separate settings.py file). It should be okay for now but create an issue to track it.
2. Please update the documentation on how the Database objects are created and the steps to take when some one wants to connect to a new Database.

**Comment by mathemancer** [comment]
> I'll let @silentninja take a full look at this, I just skimmed through the code and added some comments.
> 
> I'm wondering if we need the DB-per-session functionality. Could we maintain a list of DBs in Ansible or in some script elsewhere and figure out which DB to send people to _outside_ of the product? I'm mainly trying to avoid building a bunch of live demo related middleware or conditionals directly in the product code.
> 
> If this will take too long to figure out, we can create an issue for it instead of fixing it now.

I did consider avoiding changing the Mathesar service itself, but went with the current approach for the following reasons:
- speed of implementation and overall simplicity (biggest consideration)
- I couldn't figure out how to avoid some branching on `LIVE_DEMO` mode in the web service even if another abstraction were added (assuming one webservice; this would be avoidable by also having multiple web service installations).
- Most of the changes in this PR are actually general improvements aside from live demo concerns:
  - installation improvement
  - avoiding re-reading `config/setttings.py` on the running service.
- The _only_ changes which would be reverted later to completely take out live-demo specifics be:
  - remove the new `LiveDemoMiddleware`
  - remove the `demo` directory
  - take the conditional out of the initial page load template view.
  The last is the one I'm least comfortable with.

All that said, I actually think the Live Demo Mode should stay in the product long-term, since many data products come with a way to load sample data sets and play with them, and this enables that.

Given the feedback, I'm going to make a few changes, and re-request review from @silentninja. (You're welcome to re-review as well, @kgodey if you have time).

File: db/install.py
Comment (kgodey) [reply db/install.py:19]: Okay please mark #1481 a draft until you've updated it, @silentninja.

**Comment by mathemancer** [comment]
Okay, the new version of this PR has the following changes from when you last reviewed:

- Moved the demo-specific middleware to `demo.middleware`
- Moved demo-specific settings (adding the middleware) to `demo.settings`
- Got rid of the `LIVE_DEMO` env var. Replaced it with using the `DJANGO_SETTINGS_MODULE` variable (more idiomatic, I think).

Now,  to get rid of all demo-specific logic, one need only delete the `demo` directory.

To _use_ the demo-specific logic, one should set `DJANGO_SETTINGS_MODULE=demo.settings` in their `.env` file.

**Comment by silentninja** [comment]
Looks good, I don't have any concerns now.

**Review by silentninja — APPROVED** [review]

File: demo/settings.py
Comment (mathemancer) [reply demo/settings.py:3]: I'll go ahead and add it here. It'll have to be the caller's responsibility to handle the case where the variable is missing (i.e., when it's not `False`).

**Pull request merged**
